RFM9x.nut

```diff
@@ -0,0 +1,399 @@
+    static REG_TCXO = 0x4B;
+    
+    // -------------------- Some Tables --------------------------- //
+    
+    static BWTABLE = {
+        "7.8" : 0x00,
+        "10.4" : 0x01,
+        "15.6" : 0x02,
+        "20.8" : 0x03,
+        "31.25": 0x04,
+        "41.7": 0x05,
+        "62.5": 0x06,
+        "125": 0x07,
+        "250" : 0x08,
+        "500": 0x09
+    };
+    
+    static CRTABLE = {
+        "4/5": 0x01,
+        "4/6": 0x02,
+        "4/7": 0x03,
+        "4/8": 0x04
+    };
+    
+    static SLEEP = 0x00;
+    static STANDBY = 0x01;
+    static FSTX = 0x02;
+    static TX = 0x03;
+    static FSRX = 0x04;
+    static RXCONTINUOUS = 0x05;
+    static RXSINGLE = 0x06;
+    static CAD = 0x07;
+    
+    static _txHeaderTo = 0xFF;
+    static _txHeaderFrom = 0x00;
+    static _txHeaderId = 0;
+    static _thisAddress = 0x66;
+    static _txHeaderFlags = 0;
+    
+    constructor(spi, irq, cs=null) {
+        // Assume spi is already initialized, assume cs is already initialized
+        this._spi = spi;
+        this._irqPin = irq;
+        this._cs = cs;
+    }
+    
+    function init() {
+        setMode(SLEEP); // need to do this to go into LoRa mode
+        
+        // Defaults
+        setFrequency(915000000);
+        setBandwidth("125");
+        setSpreadingFactor(7);
+        setCodingRate("4/5");
+        setImplicitHeaderMode(false);
+        setPreambleLength(8);
+        setPayloadLength(10);
+
+        configIRQ();
+        
+        _setInLoRaMode();
+
+    }
+
+    function configIRQ() {
+        if(_irqPin) {
+            _irqPin.configure(DIGITAL_IN_PULLDOWN, _isr.bindenv(this));
+        }
+        else {
+            throw "interrupt pin passed to constructor is invalid";
+        }
+    }
+
+    function _isr() {
+        // Active high]
+        if(_irqPin.read()) {
+            local read = readReg(REG_IRQ_FLAGS);
+            if (read & (1 << PAYLOAD_CRC_ERROR)) {
+                _error = true;
+            }
+            else if(read & (1 << TX_DONE)) {
+                _sendFinished = true;
+            }
+            else if (read == ((1 << RX_DONE) | (1 << VALID_HEADER))) {
+                _receive(_readFromRXBuffer());
+            }
+            
+            clearInterrupts();
+        }
+    }
+
+    function _receive(data) {
+        if(_receiveHandler) {
+            _receiveHandler(data);
+        }
+    }
+    
+    function _writeToTXBuffer(data, len) {
+        //writeReg(REG_FIFO_TX_BASE_ADDR, 0x00);
+        // writeReg(REG_FIFO_ADDR_PTR, 0x00);
+        local write = blob(len + 1);
+        write[0] = (FIFO | WRITE_MASK);
+        for(local i = 0; i < len; ++i) {
+            //writeReg(FIFO,data[i]);
+            write[i+1] = data[i];
+        }
+        _csLow();
+        _spi.writeread(write);
+        _csHigh();
+    }
+    
+    function sendData(data, header, footer) {
+        local len = data.len();
+
+        local extra_len = 0;
+        extra_len += (header ? 4 : 0);
```

Comment:
This magic numbers should either be explained in the comments or in a mnemonic constants used for them.